### PR TITLE
Convert Teradata functions to AbstractTestFunctions

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-teradata-functions/src/test/java/com/facebook/presto/teradata/functions/TestTeradataDateFunctions.java [hunk 1]
```diff
@@ -121,33 +135,34 @@ public void testEscapedText()
...
+    private static SqlTimestamp toTimestamp(DateTime dateTime)
     {
         return new SqlTimestamp(dateTime.getMillis(), SESSION.getTimeZoneKey());
     }
 
+    @SuppressWarnings("SameParameterValue")
     private void assertTimestamp(String projection, int year, int month, int day, int hour, int minutes, int seconds)
     {
-        FUNCTION_ASSERTIONS.assertFunction(
+        functionAssertions.assertFunction(
```
Comment (kokosing) [inline presto-teradata-functions/src/test/java/com/facebook/presto/teradata/functions/TestTeradataDateFunctions.java]: you can just call `this.assertFunction()` instead

**Comment by martint** [comment]
> Unfortunately poor testng does not super anything like `@Rule`

Oooh, I should check that. If those provide the semantics we want, we should consider migrating to junit.

File: presto-teradata-functions/src/test/java/com/facebook/presto/teradata/functions/TestTeradataDateFunctions.java (hunk 1 above)
Comment (electrum) [inline presto-teradata-functions/src/test/java/com/facebook/presto/teradata/functions/TestTeradataDateFunctions.java]: Thanks, all these can actually just be `assertFunction()` (doesn't even need `this`)

**Comment by electrum** [comment]
@kokosing I agree, though using `@Rule` here would be more verbose, as you'd need to prefix the call:
```java
@Rule FunctionAssertions assertions = ...;

@Test
void test() {
    assertions.assertFunction(...);
}
```
Maybe if we were ok with using a short variable name:
```java
@Rule FunctionAssertions x = ...;

@Test
void test() {
    x.assertFunction(...);
}
```

**Comment by martint** [comment]
It turns out `@Rule` doesn't do what I need. But! I managed to coerce junit 4 do what I wanted, and I now have working prototype of a new test runner that can manage state objects in a similar way to how JMH does it. You can tag state objects as suite/class/method-scoped and declare field or argument dependencies in your test classes. The framework takes care of instantiating them, injecting them into fields and test methods, and tearing them down at the right time.

In case you're interested: https://github.com/martint/presto/commit/ef7f759223fdc7f654f9fa51a819f680c512a2b1

**Comment by kokosing** [comment]
Nice! Would it be able to inject different implementations for injecting points (like different `QueryRunner`s)?

Have you seen junit 5? They refactored rules and runners with extension framework which supports extensions by annotation, parameter resolving (as you just implemented) or conditional test execution etc. See http://blog.codefx.org/design/architecture/junit-5-extension-model/ and http://junit.org/junit5/docs/current/user-guide/#extensions. It seems to be powerful.

CC: @ArturGajowy

**Comment by martint** [comment]
I hadn't seen junit 5. It seems like it does or it should be possible to extend it to do what we need. I wonder if it's well-supported by intellij and surefire yet.

**Comment by kokosing** [comment]
I think so, even if not it will get this support really soon. Junit is a widely used library, I think it is much more popular than testng.

Also I think we should write down all the problems we have with the current test framework. So we could introduce a new testing framework in case of possible tests migration. I am going to create an issue for that.

**Comment by kokosing** [comment]
See https://github.com/prestodb/presto/issues/6910

**Comment by ArturGajowy** [comment]
@martint the JUnit Team cooperates very closely with JetBrains. JetBrains even [released a hotfix](https://blog.jetbrains.com/idea/2016/12/intellij-idea-2016-3-1-rc-updates-junit-5-support-to-m3/) for a recent problem with JUnit support.

Regarding surefire - here's a [sample project with mvn + junit5](https://github.com/junit-team/junit5-samples/tree/master/junit5-maven-consumer). The surefire support is offered by JUnit Team itself, as a [surefire provider](https://github.com/junit-team/junit5-samples/blob/master/junit5-maven-consumer/pom.xml#L44).

**Pull request closed without merging**